### PR TITLE
[JENKINS-39402] Cap the number of group headers printed by AccessDeniedException2

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-39402](https://issues.jenkins-ci.org/browse/JENKINS-39402)

@reviewbybees

### CODE REVIEW

File: test/src/test/java/hudson/security/AccessDeniedException2Test.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            groups[i] = "group" + i;
+        }
+        realm.addGroups("user", groups);
+        r.jenkins.setSecurityRealm(realm);
+        r.jenkins.setAuthorizationStrategy(new MockAuthorizationStrategy());
+        try {
+            r.createWebClient().login("user").goTo("confgure");
+            fail("should not have been allowed to access anything");
+        } catch (FailingHttpStatusCodeException x) {
+            assertEquals(HttpURLConnection.HTTP_FORBIDDEN, x.getStatusCode());
```
Comment (jglick) [inline test/src/test/java/hudson/security/AccessDeniedException2Test.java]: Without fix, fails here with a 500 from Jetty as reported in the field.

File: test/src/test/java/hudson/security/AccessDeniedException2Test.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            List<String> reportedGroups = new ArrayList<>();
+            for (NameValuePair header : x.getResponse().getResponseHeaders()) {
+                if (header.getName().equals("X-You-Are-In-Group")) {
+                    reportedGroups.add(header.getValue());
+                }
+            }
+            assertThat("capped at a reasonable number", reportedGroups, Matchers.<List<String>>allOf(
+                Matchers.<String>hasSize(11), // 10 groups plus final warning
+                Matchers.<String>hasItem("<991 more>"))); // 1000 + SecurityRealm.AUTHENTICATED_AUTHORITY.getAuthority() - 10
```
Comment (jglick) [inline test/src/test/java/hudson/security/AccessDeniedException2Test.java]: order may be arbitrary

File: core/src/main/java/hudson/security/AccessDeniedException2.java [hunk 3]
```diff
@@ -12,6 +12,9 @@
  * @author Kohsuke Kawaguchi
  */
 public class AccessDeniedException2 extends AccessDeniedException {
+
+    private static final int MAX_REPORTED_AUTHORITIES = 10;
```
Comment (daniel-beck) [inline core/src/main/java/hudson/security/AccessDeniedException2.java]: Seems to be too low to be useful, especially given crazy behaviors such as https://github.com/jenkinsci/ldap-plugin/blob/4ced65b080ff3eb81ad23916ccb6133077b80185/src/main/java/hudson/security/LDAPSecurityRealm.java#L1039

**Review by daniel-beck — APPROVED** [review]
Looks good, although I'm not convinced that with the low limit any useful behavior in nontrivial security realms (where this is actually needed) will remain.

**Comment by rsandell** [comment]
:bee:

**Comment by jglick** [comment]
> any useful behavior in nontrivial security realms

I think this is more a place to start diagnosis. You can always go to `/whoAmI` etc. to see the gory details. I do not want to print many headers for fear of pushing against limits in haproxy and the like.

**Comment by jglick** [comment]
@reviewbybees done

**Comment by daniel-beck** [comment]
Should this not rather count both headers and individual authority name length as well? I.e. cap individual names at ~50 chars long or something?

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
I agree with @daniel-beck that the limit seems to be too small. In order to avoid negative, the limit should be modifiable via `SystemProperties` at least

**Comment by jglick** [comment]
These suggestions are more effort than they are worth. I would rather just delete the functionality altogether.

File: core/src/main/java/hudson/security/AccessDeniedException2.java [hunk 4]
```diff
@@ -38,8 +41,14 @@ public AccessDeniedException2(Throwable t, Authentication authentication, Permis
      */
     public void reportAsHeaders(HttpServletResponse rsp) {
         rsp.addHeader("X-You-Are-Authenticated-As",authentication.getName());
-        for (GrantedAuthority auth : authentication.getAuthorities()) {
-            rsp.addHeader("X-You-Are-In-Group",auth.getAuthority());
+        GrantedAuthority[] authorities = authentication.getAuthorities();
```
Comment (jtnord) [inline core/src/main/java/hudson/security/AccessDeniedException2.java]: do we even need to report this given there is a `whoAmI` which reports this in a much nicer way (and most users won't check the headers, and most scripts can't take any action)

File: core/src/main/java/hudson/security/AccessDeniedException2.java (hunk 3 above)
Comment (jtnord) [inline core/src/main/java/hudson/security/AccessDeniedException2.java]: tbh I'm not sure reporting this is ever useful, given the `whoAmI` url

**Comment by jtnord** [comment]
> These suggestions are more effort than they are worth. I would rather just delete the functionality altogether.

I do not see a use for this funcationality given there is a `whoAmI` so +1 to that.  but :bee: as it stands as it fixes the issue

**Review by jtnord — APPROVED** [review]
Fixes the issue - but not sure it makes sense to leave something capped in the code that you can get from `whoAmI` with a full list.

**Comment by oleg-nenashev** [comment]
I would rather like to see a test demonstrating the correct behavior in both modes than just removal of the test logic for the disabled logic

**Comment by jglick** [comment]
> a test demonstrating the correct behavior in both modes

If it is disabled by default, who cares? There was no test before and patch merely conditionally runs the original code. The test logic in the initial version of the patch was to verify that the more complicated capping logic worked.

File: core/src/main/java/hudson/security/AccessDeniedException2.java [hunk 5]
```diff
@@ -38,8 +43,10 @@ public AccessDeniedException2(Throwable t, Authentication authentication, Permis
      */
     public void reportAsHeaders(HttpServletResponse rsp) {
         rsp.addHeader("X-You-Are-Authenticated-As",authentication.getName());
-        for (GrantedAuthority auth : authentication.getAuthorities()) {
-            rsp.addHeader("X-You-Are-In-Group",auth.getAuthority());
+        if (REPORT_GROUP_HEADERS) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/security/AccessDeniedException2.java]: If false, I would rather return a message like "Disabled due to JENKINS-39402; use /whoAmI to diagnose groups. Or set the XXX system property to true" within the Header value

**Review by jtnord — APPROVED** [review]
:bee:

**Pull request merged**